### PR TITLE
update loss functions - siou loss is added 

### PR DESCRIPTION
**Pull request by diawmous123**
The SIou Loss has been  added to losses.py. You have to use either the focal loss or not  by enabling or desabling the focal_loss option in losses.py

### CODE REVIEW

**Comment by ibrandiay** [comment]
@diawmous123 please fix the branch conficts with main branch

File: yolo_one/inference.py [hunk 1]
```diff
@@ -15,7 +15,7 @@
 import time
 from pathlib import Path
 import logging
-
+import rerun as rr
```
Comment (ibrandiay) [inline yolo_one/inference.py]: rerun should be added in another branch workflow

File: yolo_one/losses.py [hunk 2]
```diff
@@ -357,10 +366,15 @@ def _ciou_loss(self, pred_boxes: torch.Tensor, target_boxes: torch.Tensor) -> to
...
+        # CIoU Loss
         ciou = iou - rho2 / torch.clamp(enclose_c2, min=1e-6) - alpha * v
+
+        if self.focal_loss:
+            Loss = (iou**self.focal_gamma)*(1 - ciou)
+        else:
+            Loss = 1 - ciou
         
-        return 1 - ciou
+        return Loss
```
Comment (ibrandiay) [inline yolo_one/losses.py]: Variable names must follow the snake_case naming convention
https://peps.python.org/pep-0008/

File: yolo_one/losses.py [hunk 3]
```diff
@@ -395,64 +409,157 @@ def _eiou_loss(
...
         eiou = iou - rho2 / (c2+ 1e-6) - rho2_w / (wc2 + 1e-6) - rho2_h / (hc2 + 1e-6)
-        return (iou**self.focal_gamma)*(1 - eiou)
+        
+        if self.focal_loss:
+            Loss = (iou**self.focal_gamma)*(1 - eiou)
+        else:
+            Loss = 1 - eiou
+
+        return Loss
+
```
Comment (ibrandiay) [inline yolo_one/losses.py]: Variable names must follow the snake_case naming convention
https://peps.python.org/pep-0008/

File: yolo_one/losses.py [hunk 4]
```diff
@@ -485,5 +594,12 @@ def create_yolo_one_loss(
         focal_gamma=focal_gamma,
         iou_type=iou_type,
         label_smoothing=label_smoothing,
+<<<<<<< HEAD
+        p5_weight_boost=p5_weight_boost,
+        theta = theta,
+        focal_loss=focal_loss
+    )
+=======
```
Comment (ibrandiay) [inline yolo_one/losses.py]: Please run the codes before submitting them, because this piece of code does not even run given the erroneous characters in the code: <<<<<<< HEAD"

File: yolo_one/losses.py [hunk 5]
```diff
@@ -485,5 +594,12 @@ def create_yolo_one_loss(
...
         p5_weight_boost=p5_weight_boost
     )
+>>>>>>> 9c24f5a62b2e62de27cee5aac94701ed4c47af96
```
Comment (ibrandiay) [inline yolo_one/losses.py]: ?

File: yolo_one/losses.py (hunk 4 above)
Comment (ibrandiay) [inline yolo_one/losses.py]: ?

**Comment by diawmous123** [comment]
@ibrandiay It is done. Please, let me know if it is ok?

File: yolo_one/losses.py [hunk 6]
```diff
@@ -10,7 +10,7 @@
 import torch.nn.functional as F
 from yolo_one.utils.general import box_cxcywh_to_xyxy
 from typing import List, Tuple, Dict, Optional
-import numpy as np
+from termcolor import colored
```
Comment (ibrandiay) [inline yolo_one/losses.py]: **termcolor** is not relevant here, in addition you should always favor native python packages

File: yolo_one/losses.py [hunk 7]
```diff
@@ -517,7 +516,7 @@ def _meiou_loss(
 
         enclosed_x1, enclosed_y1 = torch.min(pred_x1, target_x1), torch.min(pred_y1, target_y1)
         enclosed_x2, enclosed_y2 = torch.max(pred_x2, target_x2), torch.max(pred_y2, target_y2)
-        enclosed_w, enclosed_h = enclosed_x2 - enclosed_x1, enclosed_y2 - enclosed_y1
+        enclosed_w, enclosed_h = torch.abs(enclosed_x2 - enclosed_x1), torch.abs(enclosed_y2 - enclosed_y1)
```
Comment (ibrandiay) [inline yolo_one/losses.py]: why abs here ?

File: tests/unit/test_losses.py [hunk 8]
```diff
@@ -150,6 +150,19 @@ def test_shape_conf_loss(self):
...
+        pred_box = None
+        target_box = torch.tensor([[1.0, 1.0, 2.0, 2.0]])
+
+        for name in self.loss_methods:
+
+            loss_fn = getattr(self.obj, name)
+            loss = loss_fn(pred_box, target_box)
+            self.assertIsInstance(loss, torch.Tensor)
+            self.assertEqual(loss.item(), 0.0)  """
+
```
Comment (ibrandiay) [inline tests/unit/test_losses.py]: Who is doing this code block put in docstring here? What is its relevance?

File: tests/unit/test_losses.py [hunk 9]
```diff
@@ -166,18 +179,19 @@ def test_idententical_boxes(self):
...
-        pred_box = torch.tensor([[0.0, 0.0, 1.0, 1.0]])
-        target_box = torch.tensor([[1.0, 1.0, 2.0, 2.0]])
+
+        pred_box = torch.tensor([0.0, 1.0, 3.0, 3.0])
+        target_box = torch.tensor([0.0, 0.0, 4.0, 4.0])
         for name in self.loss_methods:
             loss_fn = getattr(self.obj, name)
             loss = loss_fn(pred_box, target_box)
-            self.assertGreater(loss, 0, f"{name}: loss should be > 0")
+            self.assertGreaterEqual(loss, 0, f"{name}: loss should be > 0")
```
Comment (ibrandiay) [inline tests/unit/test_losses.py:230]: what code format do you use here for assertGreaterEqual ?

File: tests/unit/test_losses.py [hunk 10]
```diff
@@ -215,18 +229,18 @@ def test_invalid_flipped_boxes(self):
...
+    #     test_cases = [
+    #         (torch.tensor([[float("nan"), 0.0, 1.0, 2.0]]), torch.tensor([[0.0, 0.0, 1.0, 2.0]])),
+    #         (torch.tensor([[0.0, 0.0, 1.0, 2.0]]), torch.tensor([[0.0, 0.0, float("inf"), 2.0]])),
+    #         (torch.tensor([[float("-inf"), 0.0, 1.0, 2.0]]), torch.tensor([[0.0, 0.0, 1.0, 2.0]])),
+    #     ]
+    #     for name in self.loss_methods:
+    #         loss_fn = getattr(self.obj, name)
+    #         for pred_box, target_box in test_cases:
+    #             loss = loss_fn(pred_box, target_box)
+    #         self.assertTrue(torch.isfinite(loss), f"{name} must handle NaN/ Inf inputs (got {loss})")
```
Comment (ibrandiay) [inline tests/unit/test_losses.py:274]: it possible to remove this block ?

File: yolo_one/losses.py [hunk 11]
```diff
@@ -314,6 +315,30 @@ def _compute_objectness_loss(
...
         """Complete IoU loss implementation for anchor-free"""
+
+
+        # Pred and target boxes do not contain None 
+        assert pred_boxes is not None, "pred_boxes must not be None"
+        assert target_boxes is not None, "target_boxes must not be None"
+
+        # Pred and target boxes do not contain NaN or Inf 
+        assert torch.isfinite(pred_boxes).all(), "pred_boxes contains NaN or Inf"
+        assert torch.isfinite(target_boxes).all(), "target_boxes contains NaN or Inf"
```
Comment (ibrandiay) [inline yolo_one/losses.py]: How can a target contain an aN or an inf? As well as the output of a network? Assertions on None are more than enough, this is incorrect

File: yolo_one/losses.py [hunk 12]
```diff
@@ -314,6 +315,30 @@ def _compute_objectness_loss(
...
+
+        # Pred and target boxes must be Tensor
+        assert isinstance(pred_boxes, torch.Tensor), f"pred_boxes must be a torch.Tensor, got {type(pred_boxes)}"
+        assert isinstance(target_boxes, torch.Tensor), f"target_boxes must be a torch.Tensor, got {type(target_boxes)}"
+
+        # The type of pred and target must be float tensor
+        assert pred_boxes.dtype.is_floating_point, f"pred_boxes must be a float tensor, got {pred_boxes.dtype}"
```
Comment (ibrandiay) [inline yolo_one/losses.py]: it cant be half precision ( float16)  or torch.cuda.FloatTensor...tensor in gpu can't pass this assertion , please teste your code in gpu before add it

File: yolo_one/losses.py [hunk 13]
```diff
@@ -314,6 +315,30 @@ def _compute_objectness_loss(
...
+        assert target_boxes.dtype.is_floating_point, f"target_boxes must be a float tensor, got {target_boxes.dtype}"
+
+        # Pred and target boxes must have the same shape
+        assert pred_boxes.shape == target_boxes.shape, (
```
Comment (ibrandiay) [inline yolo_one/losses.py]: channel can be different

File: yolo_one/losses.py [hunk 14]
```diff
@@ -383,6 +408,31 @@ def _eiou_loss(
...
         
+        """Efficient IoU loss implementation for anchor-free"""
+        
+        # Pred and target boxes do not contain None 
+        assert pred_boxes is not None, "pred_boxes must not be None"
+        assert target_boxes is not None, "target_boxes must not be None"
+
+        # Pred and target boxes do not contain NaN or Inf 
+        assert torch.isfinite(pred_boxes).all(), "pred_boxes contains NaN or Inf"
+        assert torch.isfinite(target_boxes).all(), "target_boxes contains NaN or Inf"
```
Comment (ibrandiay) [inline yolo_one/losses.py]: How can a target contain an aN or an inf? As well as the output of a network? Assertions on None are more than enough, this is incorrect

File: yolo_one/losses.py [hunk 15]
```diff
@@ -383,6 +408,31 @@ def _eiou_loss(
...
+
+        # Pred and target boxes must be Tensor
+        assert isinstance(pred_boxes, torch.Tensor), f"pred_boxes must be a torch.Tensor, got {type(pred_boxes)}"
```
Comment (ibrandiay) [inline yolo_one/losses.py]: and torch.cuda.FloatTensor ?

File: yolo_one/losses.py [hunk 16]
```diff
@@ -383,6 +408,31 @@ def _eiou_loss(
...
+        assert isinstance(target_boxes, torch.Tensor), f"target_boxes must be a torch.Tensor, got {type(target_boxes)}"
+
+        # The type of pred and target must be float tensor
+        assert pred_boxes.dtype.is_floating_point, f"pred_boxes must be a float tensor, got {pred_boxes.dtype}"
+        assert target_boxes.dtype.is_floating_point, f"target_boxes must be a float tensor, got {target_boxes.dtype}"
+
+        # Pred and target boxes must have the same shape
+        assert pred_boxes.shape == target_boxes.shape, (
+            f"pred_boxes and target_boxes must have the same shape, "
```
Comment (ibrandiay) [inline yolo_one/losses.py]: shap can be diff

File: yolo_one/losses.py [hunk 17]
```diff
@@ -412,14 +462,11 @@ def _eiou_loss(
...
-        eiou = iou - rho2 / (c2+ 1e-6) - rho2_w / (wc2 + 1e-6) - rho2_h / (hc2 + 1e-6)
+        eiou_loss = 1  - iou + rho2 / c2 + rho2_w / wc2 + rho2_h / hc2 
         
         if self.focal_loss:
-            eiou_loss = (iou**self.focal_gamma)*(1 - eiou)
-        else:
-            eiou_loss = 1 - eiou
+            eiou_loss = (iou**self.focal_gamma)*(eiou_loss)
 
-        eiou_loss = torch.nan_to_num(eiou_loss, nan=0.0, posinf=0.0, neginf=0.0)
```
Comment (ibrandiay) [inline yolo_one/losses.py]: don't make sense

File: yolo_one/losses.py [hunk 18]
```diff
@@ -498,7 +565,30 @@ def _siou_loss(
...
- 
+        
+        """SIoU loss implementation for anchor-free"""
+
+        # Pred and target boxes do not contain None 
+        assert pred_boxes is not None, "pred_boxes must not be None"
+        assert target_boxes is not None, "target_boxes must not be None"
+
+        # Pred and target boxes do not contain NaN or Inf 
+        assert torch.isfinite(pred_boxes).all(), "pred_boxes contains NaN or Inf"
```
Comment (ibrandiay) [inline yolo_one/losses.py]: review it

File: yolo_one/losses.py [hunk 19]
```diff
@@ -498,7 +565,30 @@ def _siou_loss(
...
+        assert torch.isfinite(target_boxes).all(), "target_boxes contains NaN or Inf"
+
+        # Pred and target boxes must be Tensor
+        assert isinstance(pred_boxes, torch.Tensor), f"pred_boxes must be a torch.Tensor, got {type(pred_boxes)}"
```
Comment (ibrandiay) [inline yolo_one/losses.py]: !!!

File: yolo_one/losses.py [hunk 20]
```diff
@@ -498,7 +565,30 @@ def _siou_loss(
...
+        assert isinstance(target_boxes, torch.Tensor), f"target_boxes must be a torch.Tensor, got {type(target_boxes)}"
+
+        # The type of pred and target must be float tensor
+        assert pred_boxes.dtype.is_floating_point, f"pred_boxes must be a float tensor, got {pred_boxes.dtype}"
+        assert target_boxes.dtype.is_floating_point, f"target_boxes must be a float tensor, got {target_boxes.dtype}"
+
+        # Pred and target boxes must have the same shape
+        assert pred_boxes.shape == target_boxes.shape, (
```
Comment (ibrandiay) [inline yolo_one/losses.py]: !!!

File: yolo_one/losses.py (hunk 19 above)
Comment (diawmous123) [reply yolo_one/losses.py]: If the instance is "torch.cuda.FloatTensor", the code will only run on the CPU. And if the instance is  "torch.Tensor", the code will run on both the CPU and the GPU. torch.Tensor is more general

File: tests/unit/test_losses.py (hunk 10 above)
Comment (diawmous123) [reply tests/unit/test_losses.py:274]: Yes, I will do it

File: yolo_one/losses.py (hunk 11 above)
Comment (diawmous123) [reply yolo_one/losses.py]: It will be removed from the loss function

File: yolo_one/losses.py (hunk 12 above)
Comment (diawmous123) [reply yolo_one/losses.py]: If the instance is torch.cuda.FloatTensor, the code will only run on the GPU. If the instance is torch.Tensor, the code can run on both CPU and GPU. torch.Tensor is more general. I tested it in the GPU and it works fine  !!!

File: yolo_one/losses.py (hunk 13 above)
Comment (diawmous123) [reply yolo_one/losses.py]: It will be fixed !! Thanks !

File: yolo_one/losses.py (hunk 15 above)
Comment (diawmous123) [reply yolo_one/losses.py]: As said previously, torch.Tensor works fine both in the GPU and CPU. It is more generale than torch.cuda.FloatTensor. However, i can force it to run only on the GPU

File: yolo_one/losses.py (hunk 16 above)
Comment (diawmous123) [reply yolo_one/losses.py]: Will be fixed. Thanks !!!

File: yolo_one/losses.py (hunk 17 above)
Comment (diawmous123) [reply yolo_one/losses.py]: It is removed in the new version. We don't need it.

File: yolo_one/losses.py (hunk 18 above)
Comment (diawmous123) [reply yolo_one/losses.py]: It will be removed

File: yolo_one/losses.py (hunk 20 above)
Comment (diawmous123) [reply yolo_one/losses.py]: see above

**Comment by diawmous123** [comment]
All these remarks will be addressed in the new version. Thank you for the review !

**Comment by diawmous123** [comment]
I pushed the new version of the losses and the test functions to the Git repository. Please, let me know if you have any comments. Thanks !

File: yolo_one/losses.py [hunk 21]
```diff
@@ -313,70 +314,52 @@ def _compute_objectness_loss(
...
-        assert torch.isfinite(target_boxes).all(), "target_boxes contains NaN or Inf"
 
         # Pred and target boxes must be Tensor
-        assert isinstance(pred_boxes, torch.Tensor), f"pred_boxes must be a torch.Tensor, got {type(pred_boxes)}"
-        assert isinstance(target_boxes, torch.Tensor), f"target_boxes must be a torch.Tensor, got {type(target_boxes)}"
-
-        # The type of pred and target must be float tensor
-        assert pred_boxes.dtype.is_floating_point, f"pred_boxes must be a float tensor, got {pred_boxes.dtype}"
-        assert target_boxes.dtype.is_floating_point, f"target_boxes must be a float tensor, got {target_boxes.dtype}"
+        assert isinstance(pred_boxes, torch.cuda.FloatTensor), f"pred_boxes must be a torch.cuda.FloatTensor type, got {type(pred_boxes)}"
```
Comment (ibrandiay) [inline yolo_one/losses.py]: This type is not general it does not work on CPU for example

File: yolo_one/losses.py [hunk 22]
```diff
@@ -483,188 +487,77 @@ def _meiou_loss(
...
```
Comment (ibrandiay) [inline yolo_one/losses.py]: not work in cpu

File: yolo_one/losses.py [hunk 23]
```diff
@@ -392,89 +375,110 @@ def _ciou_loss(self, pred_boxes: torch.Tensor, target_boxes: torch.Tensor) -> to
...
```
Comment (ibrandiay) [inline yolo_one/losses.py]: cpu is not support for this way !

File: yolo_one/losses.py (hunk 21 above)
Comment (diawmous123) [reply yolo_one/losses.py]: Yes exactly ! I told you.

File: yolo_one/losses.py (hunk 22 above)
Comment (diawmous123) [reply yolo_one/losses.py]: see above

File: yolo_one/losses.py (hunk 23 above)
Comment (diawmous123) [reply yolo_one/losses.py]: See above. torch.Tensor is more general

Comment (diawmous123) [reply yolo_one/losses.py]: Fixed !

File: yolo_one/losses.py (hunk 22 above)
Comment (diawmous123) [reply yolo_one/losses.py]: Fixed !

File: yolo_one/losses.py (hunk 21 above)
Comment (diawmous123) [reply yolo_one/losses.py]: Fixed !
